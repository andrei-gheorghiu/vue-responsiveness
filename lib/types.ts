export type VueResponsivenessBreakpoints = Record<string, number | null>;

export type VueResponsivenessMatches<T extends string = ""> = Record<
  T,
  { min: boolean; max: boolean; only: boolean }
> & { current: T };
