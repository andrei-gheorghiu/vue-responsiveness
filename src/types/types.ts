export type VueResponsivenessBreakpoints = Record<string, number | null>;

export type VueResponsivenessMatches = Record<
  string,
  { min: boolean; max: boolean; only: boolean }
>;

declare module "vue" {
  // eslint-disable-next-line no-unused-vars
  interface ComponentCustomProperties {
    $matches: VueResponsivenessMatches;
  }
}
