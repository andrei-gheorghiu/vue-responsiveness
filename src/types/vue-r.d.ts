import { VueResponsivenessMatches } from "./types";

declare module "vue" {
  // eslint-disable-next-line no-unused-vars
  interface ComponentCustomProperties {
    $matches: VueResponsivenessMatches;
  }
}
