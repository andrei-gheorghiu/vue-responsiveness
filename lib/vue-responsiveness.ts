import {
  VueResponsivenessBreakpoints,
  VueResponsivenessMatches,
  Presets,
} from "./";
import { ReactiveVariable } from "vue/macros";
import { App, computed, reactive } from "vue";

let matches: ReactiveVariable<VueResponsivenessMatches<string>>;

export const VueResponsiveness = {
  install(
    app: App,
    breakpoints: VueResponsivenessBreakpoints = Presets.Bootstrap_5
  ): App {
    const intervals: Record<string, { min: string; max: string }> =
      Object.entries(breakpoints)
        .sort(([, a], [, b]) => (a || 0) - (b || 0))
        .reduce((out, [key, min], i, arr) => {
          out[key] = {
            min: min ? `(min-width: ${min}px)` : "",
            max: arr[i + 1]?.[1]
              ? `(max-width: ${arr[i + 1]?.[1]! - 0.1}px)`
              : "",
          };
          return out;
        }, {} as Record<string, { min: string; max: string }>);
    matches = reactive({
      ...Object.assign(
        {},
        ...Object.keys(intervals).map((_) => ({
          [_]: { min: false, max: false, only: false },
        })),
        {
          current: computed(
            () =>
              Object.entries(matches).find(
                ([, value]) => typeof value !== "string" && value.only
              )?.[0] as string
          ),
        }
      ),
    }) as ReactiveVariable<VueResponsivenessMatches<keyof typeof breakpoints>>;

    Object.entries(intervals).forEach(([interval, mediaQueries]) => {
      const queryLists: Record<"min" | "max", MediaQueryList> = {
        min: window.matchMedia(mediaQueries.min),
        max: window.matchMedia(mediaQueries.max),
      };
      Object.entries(queryLists).forEach(([key, mediaQueryList]) => {
        const listener = ({ matches: val }: { matches: boolean }) => {
          const { min, max } = { ...matches[interval], [key]: val } as {
            min: boolean;
            max: boolean;
          };
          matches[interval] = { min, max, only: min && max };
        };
        mediaQueryList.addEventListener("change", listener);
        listener(mediaQueryList);
      });
    });

    app.config.globalProperties.$matches = matches;

    return app;
  },
};

export const useMatches = () => matches;
