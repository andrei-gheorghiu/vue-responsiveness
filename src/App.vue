<template>
  <div>
    <pre v-text="JSON.stringify($matches, null, 2)" />
    <h2 style="color: #fff6">
      Current interval:
      <span style="color: #bada55">{{ $matches.current }}</span>
    </h2>
  </div>
</template>
