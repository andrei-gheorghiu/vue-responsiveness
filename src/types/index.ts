export * as VueResponsiveness from "./vue-responsiveness";
export * from "./types";
