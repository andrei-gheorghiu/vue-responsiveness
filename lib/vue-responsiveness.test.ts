import { mount } from "@vue/test-utils";
import { describe, it, expect } from "vitest";
import App from "../src/App.vue";
import { VueResponsiveness, useMatches } from "./vue-responsiveness";
import { VueResponsivenessBreakpoints } from "./types";
import { Presets } from "./presets";
import { defineComponent } from "vue";

describe("vue-responsiveness", () => {
  const bs5 = Object.assign(
    {},
    {
      current: "xs",
    },
    ...Object.keys(Presets.Bootstrap_5).map((key) => ({
      [key]: {
        max: true,
        min: true,
        only: true,
      },
    }))
  );
  const render = (preset?: VueResponsivenessBreakpoints) =>
    mount(App, {
      global: {
        plugins: [[VueResponsiveness, preset]],
      },
    });

  it("should work", () => {
    expect(render().element).toMatchSnapshot();
  });

  it("should work with custom breakpoints", () => {
    expect(render({ min: null, max: 1000 }).element).toMatchSnapshot();
  });

  it("should reverse unordered values", () => {
    expect(render({ max: 1000, min: null }).element).toMatchSnapshot();
  });

  it("should match $matches", () => {
    expect(render().vm.$matches).toEqual(bs5);
  });

  it("useMatches() should work", () => {
    const App = defineComponent({
      setup() {
        const matches = useMatches();
        return { matches };
      },
      template: "<div />",
    });
    const { vm } = mount(App, {
      global: {
        plugins: [[VueResponsiveness]],
      },
    });
    expect(vm.matches).toEqual(bs5);
  });
});
