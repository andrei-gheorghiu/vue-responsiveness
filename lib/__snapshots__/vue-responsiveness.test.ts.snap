// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vue-responsiveness > should reverse unordered values 1`] = `
<div>
  <pre>
    {
  "min": {
    "min": true,
    "max": true,
    "only": true
  },
  "max": {
    "min": true,
    "max": true,
    "only": true
  },
  "current": "min"
}
  </pre>
  <h2
    style="color: rgba(255, 255, 255, 0.4);"
  >
     Current interval: 
    <span
      style="color: rgb(186, 218, 85);"
    >
      min
    </span>
  </h2>
</div>
`;

exports[`vue-responsiveness > should work 1`] = `
<div>
  <pre>
    {
  "xs": {
    "min": true,
    "max": true,
    "only": true
  },
  "sm": {
    "min": true,
    "max": true,
    "only": true
  },
  "md": {
    "min": true,
    "max": true,
    "only": true
  },
  "lg": {
    "min": true,
    "max": true,
    "only": true
  },
  "xl": {
    "min": true,
    "max": true,
    "only": true
  },
  "xxl": {
    "min": true,
    "max": true,
    "only": true
  },
  "current": "xs"
}
  </pre>
  <h2
    style="color: rgba(255, 255, 255, 0.4);"
  >
     Current interval: 
    <span
      style="color: rgb(186, 218, 85);"
    >
      xs
    </span>
  </h2>
</div>
`;

exports[`vue-responsiveness > should work with custom breakpoints 1`] = `
<div>
  <pre>
    {
  "min": {
    "min": true,
    "max": true,
    "only": true
  },
  "max": {
    "min": true,
    "max": true,
    "only": true
  },
  "current": "min"
}
  </pre>
  <h2
    style="color: rgba(255, 255, 255, 0.4);"
  >
     Current interval: 
    <span
      style="color: rgb(186, 218, 85);"
    >
      min
    </span>
  </h2>
</div>
`;
